src/providers/spatialite/qgsspatialiteprovider.cpp

```diff
@@ -4564,9 +4591,61 @@ bool QgsSpatiaLiteProvider::changeAttributeValues( const QgsChangedAttributesMap
     }
     sql += QStringLiteral( " WHERE %1=%2" ).arg( QgsSqliteUtils::quotedIdentifier( mPrimaryKey ) ).arg( fid );
 
-    ret = exec_sql( sql.toUtf8(), errMsg );
+    // prepare SQLite statement
+    sqlite3_stmt *stmt = nullptr;
+    ret = sqlite3_prepare_v2( sqliteHandle( ), sql.toUtf8().constData(), -1, &stmt, nullptr );
     if ( ret != SQLITE_OK )
     {
+      // some unexpected error occurred during preparation
+      const char *err = sqlite3_errmsg( sqliteHandle( ) );
+      errMsg = ( char * ) sqlite3_malloc( ( int ) strlen( err ) + 1 );
+      strcpy( errMsg, err );
+      handleError( sql, errMsg, savepointId );
+      return false;
+    }
+
+    // bind variables not handled directly in the string
+    QMap<int, QVariant>::iterator i;
+    for ( i = bindings.begin(); i != bindings.end(); ++i )
+    {
+      int parameter_idx = i.key();
+      const QVariant val = i.value();
+      switch ( val.type() )
+      {
+        case QVariant::ByteArray:
+        {
+          const QByteArray ba = val.toByteArray();
+          sqlite3_bind_blob( stmt, parameter_idx, ba.constData(), ba.size(), SQLITE_TRANSIENT );
+          break;
+        }
+
+        default:
+          // This will only happen if the above code is changed to bind more types,
+          // but the programmer has forgotten to handle the type here. Fatal error.
+          sqlite3_finalize( stmt );
+          throw std::runtime_error( "Attempting to bind unhandled type." );
```

Comment:
Better to Q_ASSERT( false ) here

Reply:
It seems like the Q_ASSERT is not checked in release mode, so if the programmer forgets to handle the type correctly here it could then silently cause user data loss. Since this error can never occur because of user input or data I think it's better to visibly let it bubble up to the "Caught unhandled std::exception" handler.

What do you think?